From 62057d737492f08859b20ae2980e03aa7f825c02 Mon Sep 17 00:00:00 2001
From: Yang Shi <yang.shi@windriver.com>
Date: Thu, 22 Dec 2011 16:23:05 -0800
Subject: [PATCH 169/238] MIPS: Octeon: Make IPI IRQ threading safe

Various irq handlers have been shown to not function correctly when
handled in threads. Add IRQF_NODELAY to prevent threading.

Signed-off-by: Yang Shi <yang.shi@windriver.com>
---
 arch/mips/cavium-octeon/smp.c |   10 +++++++++-
 1 files changed, 9 insertions(+), 1 deletions(-)

diff --git a/arch/mips/cavium-octeon/smp.c b/arch/mips/cavium-octeon/smp.c
index 6277c33..14cea20 100644
--- a/arch/mips/cavium-octeon/smp.c
+++ b/arch/mips/cavium-octeon/smp.c
@@ -329,6 +329,14 @@ static void octeon_init_secondary(void)
 void octeon_prepare_cpus(unsigned int max_cpus)
 {
 	u64 mask;
+	unsigned long irq_flags;
+
+#ifdef CONFIG_PREEMPT_HARDIRQS
+	irq_flags = IRQF_DISABLED | IRQF_NODELAY;
+#else
+	irq_flags = IRQF_DISABLED;
+#endif
+ 
 #ifdef CONFIG_HOTPLUG_CPU
 	unsigned long t;
 	struct linux_app_boot_info *labi;
@@ -355,7 +363,7 @@ void octeon_prepare_cpus(unsigned int max_cpus)
 	else
 		mask = 0xffff;
 	cvmx_write_csr(CVMX_CIU_MBOX_CLRX(cvmx_get_core_num()), mask);
-	if (request_irq(OCTEON_IRQ_MBOX0, mailbox_interrupt, IRQF_DISABLED,
+	if (request_irq(OCTEON_IRQ_MBOX0, mailbox_interrupt, irq_flags,
 			"mailbox0", mailbox_interrupt)) {
 		panic("Cannot request_irq(OCTEON_IRQ_MBOX0)\n");
 	}
-- 
1.7.0

